A2017/tensorflow.ipynb

```diff
@@ -0,0 +1,1357 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can verify that the gradient evaluates to 3 as expected:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 22,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "The gradient of y with respect to x is 3.0\n"
+     ]
+    }
+   ],
+   "source": [
+    "with tf.Session() as session:\n",
+    "    dy_dx_val = session.run(dy_dx)\n",
+    "    print('The gradient of y with respect '\n",
+    "          'to x is {}'.format(dy_dx_val))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "_Note: Some of you may have noticed that TensorFlow did not complain despite no value being provided for the `x` placeholder. This is because even though `x` is part of the computation graph, the derivative of `y` with respect to `x` does not involve `x`, and therefore evaluating it does not require a value to be passed for `x`._"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Exercise\n",
+    "\n",
+    "Find the minimum of the expression\n",
+    "\n",
+    "$$2(x - 2)^2 + 2(y + 3)^2$$\n",
+    "\n",
+    "using gradient descent by filling in the following code block:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 23,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "# Avoid polluting the default graph by using an alternate graph\n",
+    "with tf.Graph().as_default():\n",
+    "    tf.set_random_seed(1234)\n",
+    "\n",
+    "    # Create two scalar variables, x and y, initialized at random.\n",
+    "    # x = WRITEME.\n",
+    "    # y = WRITEME.\n",
+    "\n",
+    "    # Create a tensor z whose value represents the expression\n",
+    "    #     2(x - 2)^2 + 2(y + 3)^2\n",
+    "    # z = WRITEME.\n",
+    "    \n",
+    "    # Compute the gradients of z with respect to x and y.\n",
+    "    # dx, dy = WRITEME.\n",
+    "    \n",
+    "    # Create an assignment expression for x using the update rule\n",
+    "    #    x <- x - 0.1 * dz/dx\n",
+    "    # and do the same for y.\n",
+    "    # x_update = WRITEME.\n",
+    "    # y_update = WRITEME.\n",
+    "    \n",
+    "    with tf.Session() as session:\n",
+    "        # Run the global initializer op for x and y.\n",
+    "        # WRITEME.\n",
+    "        \n",
+    "        for _ in range(10):\n",
+    "            pass\n",
+    "            # Run the update ops for x and y.\n",
+    "            # WRITEME.\n",
+    "            \n",
+    "            # Retrieve the values for x, y, and z, and print them.\n",
+    "            # x_val, y_val, z_val = WRITEME.\n",
+    "            # print('x = {:4.2f}, y = {:4.2f}, z = {:4.2f}'.format(x_val, y_val, z_val))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Solution"
```

Comment:
If you have time to be fancy, you could put the solution in a different file to hide it, and put the magic command `%load solution/thatfile.py` in a cell instead.

Reply:
Great idea! Done via ce91418.